File: 311. Sparse Matrix Multiplication.cpp
class Solution {
public:
    // brute force: O(mnp)time, O(mn)space, 17.24%
    vector<vector<int>> multiply(vector<vector<int>>& A, vector<vector<int>>& B) {
        int m = A.size(), p = A[0].size(), n = B[0].size();
        vector<vector<int>> ret(m, vector<int>(n, 0));

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                for(int k = 0; k < p; k++){
                    if((A[i][k] != 0) && (B[k][j] != 0))
                        ret[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return ret;
    }

    // move up if check: O(mnp)time, O(mn)space, 82.24%, ok
    vector<vector<int>> multiply(vector<vector<int>>& A, vector<vector<int>>& B) {
        int m = A.size(), p = A[0].size(), n = B[0].size();
        vector<vector<int>> ret(m, vector<int>(n, 0));

        for(int i = 0; i < m; i++){
            for(int k = 0; k < p; k++){
                if(A[i][k] != 0){   // check non-zero here to speed up
                    for(int j = 0; j < n; j++){
                        if(B[k][j] != 0)
                            ret[i][j] += A[i][k] * B[k][j];
                    }
                }
            }
        }
        return ret;
    }
};

File: 311. Sparse Matrix Multiplication.py
class Solution(object):
    
    # brute force: O(mnk) time, O(1) space, TLE
    def multiply(self, A, B):
        """
        :type A: List[List[int]]
        :type B: List[List[int]]
        :rtype: List[List[int]]
        """
        m, n, k = len(A), len(A[0]), len(B[0])
        ret = [[0 for j in range(k)] for i in range(m)]
        for row in xrange(m):
            for col in xrange(k):
                for i in xrange(n):
                    ret[row][col] += A[row][i] * B[i][col]
        return ret

    # better brute force: O(mnk) time, O(1) space
    def multiply(self, A, B):
        maps = {}
        m, n, k = len(A), len(A[0]), len(B[0])
        ret = [[0 for i in xrange(k)] for _ in xrange(m)]
        for i in xrange(m):
            for j in xrange(n):
                if A[i][j] != 0:
                    for p in xrange(k):
                        ret[i][p] += A[i][j] * B[j][p]
        return ret
    
    # hash table: O(mn + nk) time, O(mn) space
    def multiply(self, A, B):
        m, n, k = len(A), len(A[0]), len(B[0])
        ret = [[0 for j in range(k)] for i in range(m)]
        nums = {}
        for row in xrange(m):
            for col in xrange(n):
                if A[row][col] != 0:
                    if col in nums:
                        nums[col][row] = A[row][col]
                    else:
                        nums[col] = {row: A[row][col]}
        print nums
        for row in xrange(n):
            for col in xrange(k):
                if B[row][col] != 0:
                    for i, val in nums.get(row, {}).items():
                        ret[i][col] += val * B[row][col] 
        return ret
        